gcloud-java-core/src/main/java/com/google/gcloud/PageImpl.java

```diff
@@ -35,6 +38,33 @@
     Page<T> nextPage();
   }
 
+  static class PageIterator<T> extends AbstractIterator<T> {
+
+    private Iterator<T> currentPageIterator;
+    private Page<T> currentPage;
+
+    PageIterator(Page<T> currentPage) {
+      this.currentPageIterator = currentPage.values().iterator();
+      this.currentPage = currentPage;
+    }
+
+    @Override
+    protected T computeNext() {
+      if (currentPageIterator.hasNext()) {
+        return currentPageIterator.next();
+      }
+      Page<T> nextPage = currentPage.nextPage();
+      if (nextPage != null) {
+        currentPage = nextPage;
+        currentPageIterator = currentPage.values().iterator();
```

Comment:
Do we really need to support such a case? 
The case with a non null page with empty values should already be impossible. Empty pages in a row seem just an error situation to me.